3.2/root/usr/share/container-scripts/mongodb/init-petset-replset.sh

```diff
@@ -0,0 +1,185 @@
+    dig "${service_name}" SRV +search +short | cut -d' ' -f4 | rev | cut -c2- | rev
+}
+
+# Performs initialization of the primary node.
+#
+# Arguments:
+# - $1 hostname
+#
+# Uses the following global variables:
+# - MONGODB_REPLICA_NAME
+# - MONGODB_ADMIN_PASSWORD
+# - WAIT_PRIMARY_STATUS
+function setup_primary() {
+    local member_host="${1}"
+
+    local endpoints
+    endpoints="$(find_endpoints)"
+
+    if [ "$(echo "${endpoints}" | grep -v "^${member_host}" | wc -l)" != '0' ]; then
+        info "Skipping replica set initialization (already exist)"
+        return
+    fi
+
+    local config="{_id: \"${MONGODB_REPLICA_NAME}\", members: [ {_id: 0, host: \"${member_host}\"} ]}"
+
+    info "Initiating MongoDB replica using: ${config}"
+    mongo admin --eval "rs.initiate(${config});${WAIT_PRIMARY_STATUS}" --quiet
+
+    info "Creating MongoDB users ..."
+    mongo_create_admin
+    mongo_create_user "-u admin -p ${MONGODB_ADMIN_PASSWORD}"
+
+    info "Successfully initialized replica set"
+}
+
+# Performs initialization of the secondary node.
+#
+# Arguments:
+# - $1 hostname
+#
+# Uses the following global variables:
+# - MONGODB_REPLICA_NAME
+# - MONGODB_ADMIN_PASSWORD
+# - FIND_PRIMARY_HOST
+# - ADD_MEMBER
+# - SLEEP_TIME
+# - WAIT_SECONDARY_STATUS
+function setup_secondary() {
+    local member_host="${1}"
+
+    local replset
+    replset="${MONGODB_REPLICA_NAME}/$(find_endpoints | paste -s -d,)"
+
+    info "Determining primary node ..."
+
+    local primary_host=''
+    local max_attempts=20
+    local attempt
+    for attempt in $(seq 1 ${max_attempts}); do
+        info "Connecting to replica set (attempt #${attempt} from ${max_attempts}) ..."
+        primary_host="$(mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --host "${replset}" --eval "${FIND_PRIMARY_HOST}" --quiet || :)"
+
+        if [ -n "${primary_host}" ]; then
+            # Unfortunately mongo client (at least in 2.4 version) outputs a lot of
+            # unrelated log messages and here we're filtering them out by
+            # finding line that starts with "PRIMARY|". We also remove port
+            # number at the end.
+            primary_host="$(echo "${primary_host}" | grep '^PRIMARY|' || :)"
+            if [ -n "${primary_host}" ]; then
+                primary_host="$(echo "${primary_host}" | cut -d'|' -f2 | cut -d':' -f1)"
+                info "Found primary node: ${primary_host}"
+                break
+            fi
+        fi
+
+        if [ ${attempt} -lt ${max_attempts} ]; then
+            info "Connection failed, waiting ${SLEEP_TIME} sec before trying again ..."
+            sleep ${SLEEP_TIME}
+            continue
+        fi
+
+        info "ERROR: couldn't determine primary node!"
+        return 1
+    done
+
+    info "Adding ${member_host} to replica set ..."
+
+    # Unfortunately we can't rely on exit code and have to inspect the output
```

Comment:
AFAIR it also were possible that connection wasn't refused but fails because user didn't exist or replica set initialization wasn't complete.